state/core_access.go

```diff
@@ -461,21 +469,26 @@ func (ca *CoreAccessor) BeginRedelegate(
 	fee Int,
 	gasLim uint64,
 ) (*TxResponse, error) {
+	signer, err := ca.getSigner(ctx)
+	if err != nil {
+		return nil, err
+	}
+
 	if amount.IsNil() || amount.Int64() <= 0 {
 		return nil, ErrInvalidAmount
 	}
 
-	from, err := ca.signer.GetSignerInfo().GetAddress()
-	if err != nil {
-		return nil, err
-	}
 	coins := sdktypes.NewCoin(app.BondDenom, amount)
-	msg := stakingtypes.NewMsgBeginRedelegate(from, srcValAddr, dstValAddr, coins)
-	signedTx, err := ca.constructSignedTx(ctx, msg, apptypes.SetGasLimit(gasLim), withFee(fee))
-	if err != nil {
-		return nil, err
+	msg := stakingtypes.NewMsgBeginRedelegate(signer.Address(), srcValAddr, dstValAddr, coins)
+	if gasLim == 0 {
+		var err error
+		gasLim, err = signer.EstimateGas(ctx, []sdktypes.Msg{msg})
+		if err != nil {
+			return nil, fmt.Errorf("estimating gas: %w", err)
+		}
```

Comment:
Note for future refactoring, I guess. The gas estimation is very repetitive and something the Signer should do automatically rather than the user before sending every transaction.
cc @cmwaters

Reply:
I guess we're still iterating on the boundary line between what is the responsibility of the signer and the responsibility of signers caller

Reply:
Pls put in issue or resolve @cmwaters @Wondertan

Reply:
@Wondertan let me know if you want gas estimation to be on the signers side

Reply:
I think it makes sense as it abstracts aways a detail that users have to do every time. If you agree with this, let's do it or park it as an app issue